contentprovider.go

```diff
@@ -163,82 +173,116 @@ func (p *contentProvider) fillMatches(ms []*candidateMatch, numContextLines int,
-			log.Panicf(
-				"%s %v infinite loop: num %d start,end %d,%d, offset %d",
-				p.id.fileName(p.idx), p.id.metaData,
-				num, lineStart, lineEnd,
-				m.byteOffset)
+			ranges = append(ranges, Range{
+				Start: Location{
+					ByteOffset: int(startOffset),
+					LineNumber: startLine,
+					Column:     utf8.RuneCount(data[startLineOffset:startOffset]) + 1,
+				},
+				End: Location{
+					ByteOffset: int(endOffset),
+					LineNumber: endLine,
+					Column:     utf8.RuneCount(data[endLineOffset:endOffset]) + 1,
+				},
+				SymbolInfo: si,
+			})
 		}
 
-		data := p.data(false)
-
-		// Due to merging matches, we may have a match that
-		// crosses a line boundary. Prevent confusion by
-		// taking lines until we pass the last match
-		for lineEnd < len(data) && endMatch > uint32(lineEnd) {
-			next := bytes.IndexByte(data[lineEnd+1:], '\n')
-			if next == -1 {
-				lineEnd = len(data)
-			} else {
-				// TODO(hanwen): test that checks "+1" part here.
-				lineEnd += next + 1
-			}
+		firstLineNumber := chunk.firstLine - numContextLines
+		if firstLineNumber < 1 {
+			firstLineNumber = 1
 		}
+		firstLineStart, _ := newlines.lineBounds(firstLineNumber)
+
+		chunkMatches = append(chunkMatches, ChunkMatch{
+			Content: newlines.getLines(data, firstLineNumber, chunk.lastLine+numContextLines+1),
+			ContentStart: Location{
+				ByteOffset: int(firstLineStart),
+				LineNumber: firstLineNumber,
+				Column:     1,
+			},
+			FileName: false,
+			Ranges:   ranges,
+		})
+	}
+	return chunkMatches
+}
 
-		finalMatch := LineMatch{
-			LineStart:  lineStart,
-			LineEnd:    lineEnd,
-			LineNumber: num,
-		}
-		finalMatch.Line = data[lineStart:lineEnd]
+type candidateChunk struct {
+	firstLine  int    // 1-based, inclusive
+	lastLine   int    // 1-based, inclusive
+	minOffset  uint32 // 0-based, inclusive
+	maxOffset  uint32 // 0-based, exclusive
+	candidates []*candidateMatch
+}
 
-		if numContextLines > 0 {
-			finalMatch.Before = p.newlines().getLines(data, num-numContextLines, num)
-			finalMatch.After = p.newlines().getLines(data, num+1, num+1+numContextLines)
+// chunkCandidates groups a set of sorted, non-overlapping candidate matches by line number. Adjacent
+// chunks will be merged if adding `numContextLines` to the beginning and end of the chunk would cause
+// it to overlap with an adjacent chunk.
+func chunkCandidates(ms []*candidateMatch, newlines newlines, numContextLines int) []candidateChunk {
+	var chunks []candidateChunk
+	for _, m := range ms {
+		startOffset := m.byteOffset
+		endOffset := m.byteOffset + m.byteMatchSz
+		firstLine, _, _ := newlines.atOffset(startOffset)
+		lastLine, _, _ := newlines.atOffset(endOffset)
+
+		// If there is no previous chunk, create a new one
+		if len(chunks) == 0 {
+			chunks = append(chunks, candidateChunk{
+				firstLine:  firstLine,
+				minOffset:  startOffset,
+				lastLine:   lastLine,
+				maxOffset:  endOffset,
+				candidates: []*candidateMatch{m},
+			})
+			continue
 		}
```

Comment:
I don't know if it is clearer, but could you somehow make the check for last below not happen if len(chunks) == 0? This block of code looks like it is the same as the else block below.

Reply:
It's slightly awkward since we need to know that `len(chunks) > 0` before getting the last chunk, but I think [the change](https://github.com/sourcegraph/zoekt/pull/380/commits/4692a8d59d15b3cea46dbe1baf3aa64eb4e82162) was a net improvement.